### ISSUE TITLE
feature request: enable herdocs by indenting with tabs instead of spaces

### ISSUE THREAD

**Issue by hojerst** [DannyBen/bashly#216]
Problem: Bashly currently wraps command script in functions and indents the code. This effectivily disables the possibility to use heredoc's. The documented suggestion is to use a script in the lib folder as this is not indented. However, this breaks again when running bashly with `-w` flag as everything is wrapped in a function again.

Suggestion: Use tabs to indent functions instead of spaces. This allows the use of `<<-EOF` heredocs which CAN be indented, but only work when text is indented with tabs instead of spaces.

**Comment by DannyBen** [comment]
How about running `unexpand -t2 your_final_script > your_tabbed_script`?

**Comment by hojerst** [comment]
As I understood it, the generated script isn't really needed for 'human consumption' but rather a artifact. Do you disagree? I mean it is beneficial to have something readable here, which consistent tab usage won't remove, nontheless this is just generated and not meant for direct editiing.

If you strongly disagree with the default indention to be tabs: how about adding a option to use tabs as a argument to the `generate` command or at least documenting your workaround above. Currently there is NO option to use heredocs when you want to use the `-w` option of bashly. This hurts much more from a readability/usability standpoint than having a slightly less readable GENERATED script.

**Comment by DannyBen** [comment]
> As I understood it, the generated script isn't really needed for 'human consumption' but rather a artifact. Do you disagree?

No. I agree.

> how about adding a option to use tabs as a argument to the `generate` command

I don't see it solving anything. 
As I see it, there are two ways to ensure the end result uses tabs instead of spaces:

1. Rewrite all of bashly's internal bash templates, to use tabs AND forcing the user to write THEIR code using tabs.
2. Post process, and just replace spaces with tabs in the generated script - which as I mentioned, can be done with one simple linux command.

> Currently there is NO option to use heredocs when you want to use the `-w` option of bashly.

How come? What is wrong with running `unexpand` on your generated script?

**Comment by hojerst** [comment]
> > how about adding a option to use tabs as a argument to the `generate` command
> 
> I don't see it solving anything. As I see it, there are two ways to ensure the end result uses tabs instead of spaces:
> 
>     1. Rewrite all of bashly's internal bash templates, to use tabs AND forcing the user to write THEIR code using tabs.

I disagree. If bashly would indent e.g. the generated `*_command` functions with a single tab instead of 2 spaces, the user could use `<<-EOF` heredocs without a problem and still use spaces for normal indention:

e.g. src/test.sh
```shell
testfunc() {
  echo whatever

  # you are only forced to use tabs within the heredoc block
  # but this is something bash itself needs to be blamed for...
  cat <<-EOF
<tab>my indented block
<tab>EOF
}

cat <<-EOF
hello
world
EOF

testfunc
```

The current behaviour of bashly breaks the non-indented hereblock in the "command script" as well as the indented one obviously.

Yes you would end up with a tab+some spaces when the original command script was using spaces for indention (which it probably will). As there will be a single tab character before each and every indented line you will still end up with consistent indention in the generated function (although the generated code mixes tabs+spaces in the same line)

>     2. Post process, and just replace spaces with tabs in the generated script - which as I mentioned, can be done with one simple linux command.

Sure. Wouldn't you agree that bashly should generate runnable code right of the box? Forcing users to use a postprocessing step just to get something runnable feels weird. I mean heredocs are a major use case imho. We are aren't talking about a matter of taste here: the generated code is just not runnable when heredocs of any form are used (with the -w option).

**Comment by DannyBen** [comment]
Obviously, we disagree on how major of a use case HEREDOC is. I do not believe it to be so.

If you need heredocs, you have the ability to use them right now, without me changing anything in bashly.

```
$ bashly generate && unexpand -t2 --first-only yourscript > yourscript.tmp && mv yourscript.tmp yourscript
```

Like in any piece of software, some things are supported out of the box, some are not. That's life.

That said - I am not opposing to the idea of solving this, either with a flag or with a more generic hook system that will run arbitrary user operations on the generated script. 

Can you just confirm that using `unexpand` on your script fixes your HEREDOC use case?

**Comment by hojerst** [comment]
> Obviously, we disagree on how major of a use case HEREDOC is. I do not believe it to be so.

Lets agree on that at least ;)

> Can you just confirm that using `unexpand` on your script fixes your HEREDOC use case?

Partly. The `--first-only` option is only implemented for the gnu version of unexpand. You either have to use the gnu version specifically or using something like `unexpand -t2,4 file` seems to do the trick to some degree. (running `unexpand -t2` will mess up the usage and other strings otherwise)

**Comment by DannyBen** [comment]
> seems to do the trick to some degree

I see. I guess my more accurate question would be: If we convert the generated script (directly in Ruby) and replacing only leading cases of two spaces with a tab character - will this work for you?

Here is my plan:

1. Add `BASHLY_TAB_INDENT` environment variable, that if set, will run this post processing on the generated script.
2. The post processing will take the output script, and run it through something like the below code to replace the leading spaces.

```ruby
def expand_tabs(tabstop = 2)
  gsub(/^( {#{tabstop}}+)/) do
    "\t" * ($1.size / tabstop)
  end
end
```

Additional notes:

1. I am leaning towards an environment variable, and not a flag, due to the fact that if someone *needs* their script to be generated with tabs, they need it *always* and not sometimes. So having a flag makes less sense, UI wise.
2. Since there are already more than a handful of environment variables, I am also considering adding support for `settings.yml` or similar, where users can configure such settings in a file that is source controlled.

WDYT?

**Comment by hojerst** [comment]
tl;dr: fine for me.

This is to some degree my original request to be honest. Although implemented differently. The script with the `<<-EOF` herdocs works completely fine when postprocessed to "leading tabs instead of spaces".

---

My thoughts regarding the additional notes (although a bit offtopic for this issue): 1. valid point 2. also. Just as a recommendation: In a fullblown solution I really like having a hierarchy of configuration settings:

1. read settings from config file `${XDG_CONFIG_HOME:-$HOME/.config}/project/config`
2. read settings from env vars
3. read settings from args

with most options configurable in all 3 "layers". Depending on the cli framework this can be implemented easily or is very complicated. But the result is very flexible from a user perspective. I personally would be completely fine with your suggested env var, though. Just my 2 cents.

**Comment by DannyBen** [comment]
> This is to some degree my original request to be honest.

Ok, then it was a misunderstanding maybe. For me it was important to differentiate between a solution that uses post processing to solve it, and one that uses tabs instead of spaces in its source. I did not want the latter.

> My thoughts regarding the additional notes

I am also a fan of hierarchical configuration, but less a fan of the `~/.config/*` configuration path.

Bashly already has a [`Settings`](https://github.com/DannyBen/bashly/blob/master/lib/bashly/settings.rb) class which gets things from the environment, so setting it to read from `./settings.yml` should be easy. I think I will avoid adding support for a central `~/.config/*` file at this point, unless you feel it is really needed?

As a side note - I am using [direnv](https://direnv.net/) - so it is completely transparent to me when I need to use environment variables as temporary configuration, on a per-directory basis.

I believe I can have all this implemented really quickly (24 hours).

**Comment by hojerst** [comment]
I'm also using `direnv` quite a lot - thats also the reason I personally don't feel the need to have a global config file.

Regarding the implementation: I already played around a little with the source. My original thought was to tweak the `Array.indent` extension to insert tabs instead of spaces if this option is on. So you might not even need a generic postprocessor. Maybe a even more generic solution would be to have a `BASHLY_INDENT` variable with `"  "` (2 spaces) being the default so you could tweak bashly to produce 3 space, 4 space or tab indents instead.

**Comment by DannyBen** [comment]
There seems to be an issue though.

```bash
testme() {
	cat <<-EOF
	this is a
	  multiline
	    heredoc text
	EOF
}

testme
```

Since the post processor replaces ALL leading spaces with tabs, the output of this would NOT be indented. Is there a way around it? Is it important?

**Comment by hojerst** [comment]
Hm you mean it would be stripping the extra indents at `multiline` and `heredoc text`? Right. This might actually be an issue when writing something like a documentation table. (like the one that is actually generated for you when calling `--help`).

How about my suggestion of not having a postprocessor but instead tweaking the `Array.indent` directly? This would lead to:

```shell
function my_command() {
	testme() {
		cat <<-EOF
		this is a
			multiline
				heredoc text
		EOF
	}

	testme
}
```

As mentioned before: this might lead to mixed indention in some cases like this:

```bash
testme() {
  cat <<-EOF
	this is a
		multiline
			heredoc text
	EOF
}

testme
```

=>

```bash
function my_command() {
	testme() {
	  cat <<-EOF
		this is a
			multiline
				heredoc text
		EOF
	}

	testme
}
```
(notice the tab followed by 2 spaces at the `cat` command).
I personally could live with that, to be honest.

**Comment by DannyBen** [comment]
> How about my suggestion of not having a postprocessor but instead tweaking the Array.indent directly?

This would solve the problem, but will create a completely mixed indentation script - since the `indent` function is only called for specific things (nested views) and not for everything.

Found one solution: Indent the lines that should stay indented using a character other than space. For example, Unicode U+3000 [　]

```bash
testme() {
	cat <<-EOF
	this is a
	　multiline
	　　heredoc text
	EOF
}

testme
```

and we are quickly going back to HEREDOC bringing more troubles than it solves... at least in bash.

**Comment by DannyBen** [comment]
@hojerst do you want to test/review the PR before I move on to the Settings stage?

**Comment by DannyBen** [comment]
@hojerst - thanks for raising this issue.

Bashly 0.7.10 is now out, with:

- `BASHLY_TAB_INDENT=1` environment variable
- Support for loading [settings from YAML](https://bashly.dannyb.co/usage/settings/)
- Additional [heredoc example](https://github.com/DannyBen/bashly/tree/master/examples/heredoc#readme)

If there is anything else you feel needs improvement, please do not hesitate to open another issue.
